microraft-store-sqlite/src/main/java/io/microraft/store/sqlite/RaftSqliteStore.java

```diff
@@ -0,0 +1,346 @@
+        dsl.close();
+    }
+
+    @Override
+    public void persistAndFlushLocalEndpoint(RaftEndpoint localEndpoint, boolean localEndpointVoting) {
+        dsl.update(KV)
+                .set(localEndpointField, localEndpoint)
+                .set(LOCAL_ENDPOINT_VOTING, localEndpointVoting)
+                .execute();
+        dsl.connection(Connection::commit);
+    }
+
+    @Override
+    public void persistAndFlushInitialGroupMembers(@Nonnull RaftGroupMembersView initialGroupMembers) {
+        dsl.update(KV).set(initialGroupMembersField, initialGroupMembers).execute();
+        dsl.connection(Connection::commit);
+    }
+
+    @Override
+    public void persistAndFlushTerm(int term, @Nullable RaftEndpoint votedFor) {
+        dsl.update(KV).set(TERM, term).set(votedForField, votedFor).execute();
+        dsl.connection(Connection::commit);
+    }
+
+    @Override
+    public void persistLogEntry(@Nonnull LogEntry logEntry) {
+        dsl.insertInto(LOG_ENTRIES, INDEX, logEntryField)
+                .values(logEntry.getIndex(), logEntry)
+                .execute();
+    }
+
+    @Override
+    public void persistSnapshotChunk(@Nonnull SnapshotChunk snapshotChunk) {
+        dsl.insertInto(SNAPSHOT_CHUNKS, INDEX, CHUNK_INDEX, chunkField)
+                .values(snapshotChunk.getIndex(), snapshotChunk.getSnapshotChunkIndex(), snapshotChunk)
+                .execute();
+        dsl.insertInto(SNAPSHOTS, INDEX, CHUNK_COUNT)
+                .values(snapshotChunk.getIndex(), snapshotChunk.getSnapshotChunkCount())
+                .onConflictDoNothing()
+                .execute();
+        tryToCleanUpOldSnapshots = true;
+    }
+
+    // Visible for testing
+    Optional<Long> getMaxCommittedSnapshotIndex() {
+        return Optional.ofNullable(
+                dsl.select(DSL.max(INDEX).as(INDEX)).from(completedSnapshots()).fetchOne(INDEX));
+    }
+
+    private static <T> Field<T> qualify(Table<?> table, Field<T> field) {
+        return DSL.field(DSL.name(table.$name(), field.$name()), field.getType());
+    }
+
+    @Override
+    public void truncateLogEntriesFrom(long logIndexInclusive) {
+        dsl.deleteFrom(LOG_ENTRIES)
+                .where(INDEX.greaterOrEqual(logIndexInclusive))
+                .execute();
+    }
+
+    @SuppressWarnings("VarUsage")
+    private Select<? extends Record1<Long>> completedSnapshots() {
+        Table<?> tempTable = DSL.table("t");
+        var blocks = dsl.select(
+                        qualify(SNAPSHOT_CHUNKS, INDEX),
+                        DSL.count(qualify(SNAPSHOT_CHUNKS, INDEX)).as(COUNT))
+                .from(SNAPSHOT_CHUNKS)
+                .groupBy(qualify(SNAPSHOT_CHUNKS, INDEX));
+        return dsl.select(qualify(tempTable, INDEX))
+                .from(blocks.asTable(tempTable), SNAPSHOTS)
+                .where(qualify(tempTable, INDEX).eq(qualify(SNAPSHOTS, INDEX)))
+                .and(DSL.field(COUNT).eq(CHUNK_COUNT));
```

Comment:
nice trick! As far as I understand this block is returning all the snapshot indices for which their all chunks are persisted. Can I ask you to write a short comment for future readers?